pkg/bloomgateway/client.go

```diff
@@ -0,0 +1,287 @@
+	for _, s := range a {
+		if !util.StringsContain(b, s) {
+			return false
+		}
+	}
+	return true
+}
+
+// listContainsAddrs checks if a slice of chunkRefAddrs contains an element
+// whos field addrs contains the same addresses as the given slice of
+// addresses.
+// It returns the index of the element, if found, and a boolean whether the
+// given list contains the given addrs.
+func listContainsAddrs(list []chunkRefsByAddrs, addrs []string) (int, bool) {
+	for i, r := range list {
+		if isEqualStringElements(r.addrs, addrs) {
+			return i, true
+		}
+	}
+	return -1, false
+}
+
+type chunkRefsByAddrs struct {
+	addrs   []string
+	refs    []*logproto.ChunkRef
+	streams []uint64
+}
+
+// groupStreamsByAddr takes a slice of stream fingerprints, a slices of chunkRef slices, and a slice of address slices
+// and groups them into a slice of chunkRefsByAddrs.
+// streams is a slice of uint64 stream fingerprints
+// chunks is a slice of chunk ref slices
+// addresses is a slice of string slices containing server addresses
+// It is necessary that len(streams) == len(chunks) == len(addresses), but the
+// function implementation does not validate the precondition and would fail silently.
+func (c *GatewayClient) groupStreamsByAddr(streams []uint64, chunks [][]*logproto.ChunkRef, addresses [][]string) []chunkRefsByAddrs {
+	res := make([]chunkRefsByAddrs, 0, len(addresses))
+	for i := 0; i < len(addresses); i++ {
+		addrs := addresses[i]
+		refs := chunks[i]
+		fp := streams[i]
+		if idx, ok := listContainsAddrs(res, addrs); ok {
+			res[idx].refs = append(res[idx].refs, refs...)
+			res[idx].streams = append(res[idx].streams, fp)
+		} else {
+			res = append(res, chunkRefsByAddrs{addrs: addrs, refs: refs, streams: []uint64{fp}})
+		}
+	}
+	return res
+}
+
+// doForAddrs sequetially calls the provided callback function fn for each
+// address in given slice addrs until the callback function does not return an
+// error.
+func (c *GatewayClient) doForAddrs(addrs []string, fn func(logproto.BloomGatewayClient) error) error {
+	var err error
+	var poolClient ringclient.PoolClient
+
+	for _, addr := range addrs {
+		poolClient, err = c.pool.GetClientFor(addr)
+		if err != nil {
+			level.Error(c.logger).Log("msg", fmt.Sprintf("failed to get client for instance %s", addr), "err", err)
+			continue
+		}
+		err = fn(poolClient.(logproto.BloomGatewayClient))
+		if err != nil {
+			level.Error(c.logger).Log("msg", fmt.Sprintf("client do failed for instance %s", addr), "err", err)
+			continue
+		}
+		return nil
+	}
+	return err
+}
+
+// serverAddrsForFingerprints returns a slices of server address slices for
+// each fingerprint of given fingerprints.
+// The indexes of the returned slices correspond to each other.
+// Returns an error in case the bloom gateway ring could not get the
+// corresponding replica set for a given fingerprint.
+func (c *GatewayClient) serverAddrsForFingerprints(tenantID string, fingerprints []uint64) ([]uint64, [][]string, error) {
```

Comment:
As we talked about earlier, it's less efficient to iterate fingerprints than tokens on large queries.

Reply:
Added a comment and log a warning in case the alogo iterating over fingerprints becomes less efficient that an algo that iterates over ring tokens and binary searches fingerprint ranges.